# Logistic regression for the Personal Loan dataset:

In [1]:
import pandas as pd
import statsmodels.api as sm

In [2]:
data = pd.read_excel("C:/Users/RAHUL REDDY/Desktop/AIML/Letsupgrade/AI-ML-July-2020-master/Day-24/drive-download-20200810T141445Z-001/Bank_Personal_Loan_Modelling.xlsx",sheet_name='Data')
data.head()

ID  Age  Experience  Income  ZIP Code  Family  CCAvg  Education  Mortgage  \
0   1   25           1      49     91107       4    1.6          1         0   
1   2   45          19      34     90089       3    1.5          1         0   
2   3   39          15      11     94720       1    1.0          1         0   
3   4   35           9     100     94112       1    2.7          2         0   
4   5   35           8      45     91330       4    1.0          2         0   

   Personal Loan  Securities Account  CD Account  Online  CreditCard  
0              0                   1           0       0           0  
1              0                   1           0       0           0  
2              0                   0           0       0           0  
3              0                   0           0       0           0  
4              0                   0           0       0           1

In [3]:
data.columns

Index(['ID', 'Age', 'Experience', 'Income', 'ZIP Code', 'Family', 'CCAvg',
       'Education', 'Mortgage', 'Personal Loan', 'Securities Account',
       'CD Account', 'Online', 'CreditCard'],
      dtype='object')

In [4]:
data.drop(['ID','ZIP Code'],axis = 1 , inplace = True)
data.head()

Age  Experience  Income  Family  CCAvg  Education  Mortgage  Personal Loan  \
0   25           1      49       4    1.6          1         0              0   
1   45          19      34       3    1.5          1         0              0   
2   39          15      11       1    1.0          1         0              0   
3   35           9     100       1    2.7          2         0              0   
4   35           8      45       4    1.0          2         0              0   

   Securities Account  CD Account  Online  CreditCard  
0                   1           0       0           0  
1                   1           0       0           0  
2                   0           0       0           0  
3                   0           0       0           0  
4                   0           0       0           1

In [5]:
y = data['Personal Loan']
X = data[['Age', 'Experience', 'Income','Family', 'CCAvg','Education', 'Mortgage','Securities Account','CD Account', 'Online', 'CreditCard']]
X1 = sm.add_constant(X)

In [6]:
logistic = sm.Logit(y,X1)

In [7]:
result = logistic.fit()

Optimization terminated successfully.
         Current function value: 0.128435
         Iterations 9


In [8]:
result.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                           Logit Regression Results                           
==============================================================================
Dep. Variable:          Personal Loan   No. Observations:                 5000
Model:                          Logit   Df Residuals:                     4988
Method:                           MLE   Df Model:                           11
Date:                Thu, 20 Aug 2020   Pseudo R-squ.:                  0.5938
Time:                        14:44:05   Log-Likelihood:                -642.18
converged:                       True   LL-Null:                       -1581.0
Covariance Type:            nonrobust   LLR p-value:                     0.000
======================================================================================
                         coef    std err          z      P>|z|      [0.025      0.975]
--------------------------------------------------------------------------------------
const                -12.1928      1.645     -7.411      0.000     -15.417      -8.968
Age                   -0.0536      0.061     -0.874      0.382      -0.174       0.067
Experience             0.0638      0.061      1.046      0.295      -0.056       0.183
Income                 0.0546      0.003     20.831      0.000       0.049       0.060
Family                 0.6958      0.074      9.364      0.000       0.550       0.841
CCAvg                  0.1240      0.040      3.127      0.002       0.046       0.202
Education              1.7362      0.115     15.088      0.000       1.511       1.962
Mortgage               0.0005      0.001      0.856      0.392      -0.001       0.002
Securities Account    -0.9368      0.286     -3.277      0.001      -1.497      -0.377
CD Account             3.8225      0.324     11.800      0.000       3.188       4.457
Online                -0.6752      0.157     -4.298      0.000      -0.983      -0.367
CreditCard            -1.1197      0.205     -5.462      0.000      -1.522      -0.718
======================================================================================
"""

__By the above result summary we can conclude that the following features (Income,Family,CCAvg,Education,Securities Account,CD Account,Online,Creditcard) are have significant impact on dependent variable(Personal Loan)__  

__To find the probability to get a Personal Loan or not, we have to calculate the 'P' value by placing the coefficient values into the formula.  
If 'P' value is > 0.5 , chance of getting Personal Loan is good.  
If 'P' value is < 0.5 , chance of getting Personal Loan is bad.__

In [9]:
import math

__For prediction we are considering the following values as independent variables with coefficient values respectivetely:__  
  
Constant - (-12.192)  
Income - 121 (0.0546)  
Family - 4 (0.695)  
CCAvg - 5.2 (0.124)  
Education - 3 (1.736)  
Securities Account - 1 (-0.936)  
CD Account - 1 (3.822)  
Online - 1 (-0.675)  
Online	CreditCard - 0 (-1.119)

In [10]:
math.exp(-(10))

4.5399929762484854e-05

__p = 1/(1+exp(-(-12.19+(0.0154*(121))+(0.695*(4))+(0.124*(5.2))+(1.73*(3))-(0.93*(1))+(3.82*(1))-(0.67*(1))-(1.119*(0)))))__

In [11]:
a = -12.19+(0.0154*(121))+(0.695*(4))+(0.124*(5.2))+(1.73*(3))-(0.93*(1))+(3.82*(1))-(0.67*(1))-(1.119*(0))

In [12]:
p = 1/(1+math.exp(-(a)))
p

0.6243844177475686

__P = 0.624  
As 'P' value is greater than 0.5, the probability of getting personal loan is more for a guy with those features.__

# Logistic regeression for Attrition dataset:

In [36]:
data1 = pd.read_csv("C:/Users/RAHUL REDDY/Desktop/AIML/Letsupgrade/AI-ML-July-2020-master/Day-24/drive-download-20200810T141445Z-001/general_data.csv")
data1.head()

Age Attrition     BusinessTravel              Department  DistanceFromHome  \
0   51        No      Travel_Rarely                   Sales                 6   
1   31       Yes  Travel_Frequently  Research & Development                10   
2   32        No  Travel_Frequently  Research & Development                17   
3   38        No         Non-Travel  Research & Development                 2   
4   32        No      Travel_Rarely  Research & Development                10   

   Education EducationField  EmployeeCount  EmployeeID  Gender  ...  \
0          2  Life Sciences              1           1  Female  ...   
1          1  Life Sciences              1           2  Female  ...   
2          4          Other              1           3    Male  ...   
3          5  Life Sciences              1           4    Male  ...   
4          1        Medical              1           5    Male  ...   

   NumCompaniesWorked Over18 PercentSalaryHike  StandardHours  \
0                 1.0      Y                11              8   
1                 0.0      Y                23              8   
2                 1.0      Y                15              8   
3                 3.0      Y                11              8   
4                 4.0      Y                12              8   

   StockOptionLevel TotalWorkingYears  TrainingTimesLastYear  YearsAtCompany  \
0                 0               1.0                      6               1   
1                 1               6.0                      3               5   
2                 3               5.0                      2               5   
3                 3              13.0                      5               8   
4                 2               9.0                      2               6   

   YearsSinceLastPromotion  YearsWithCurrManager  
0                        0                     0  
1                        1                     4  
2                        0                     3  
3                        7                     5  
4                        0                     4  

[5 rows x 24 columns]

In [37]:
data1.isna().sum()

Age                         0
Attrition                   0
BusinessTravel              0
Department                  0
DistanceFromHome            0
Education                   0
EducationField              0
EmployeeCount               0
EmployeeID                  0
Gender                      0
JobLevel                    0
JobRole                     0
MaritalStatus               0
MonthlyIncome               0
NumCompaniesWorked         19
Over18                      0
PercentSalaryHike           0
StandardHours               0
StockOptionLevel            0
TotalWorkingYears           9
TrainingTimesLastYear       0
YearsAtCompany              0
YearsSinceLastPromotion     0
YearsWithCurrManager        0
dtype: int64

In [38]:
data1['NumCompaniesWorked'].fillna(data1['NumCompaniesWorked'].mean() , inplace = True)

In [39]:
data1['TotalWorkingYears'].fillna(data1['TotalWorkingYears'].mean() , inplace = True)

In [40]:
data1.isna().sum()

Age                        0
Attrition                  0
BusinessTravel             0
Department                 0
DistanceFromHome           0
Education                  0
EducationField             0
EmployeeCount              0
EmployeeID                 0
Gender                     0
JobLevel                   0
JobRole                    0
MaritalStatus              0
MonthlyIncome              0
NumCompaniesWorked         0
Over18                     0
PercentSalaryHike          0
StandardHours              0
StockOptionLevel           0
TotalWorkingYears          0
TrainingTimesLastYear      0
YearsAtCompany             0
YearsSinceLastPromotion    0
YearsWithCurrManager       0
dtype: int64

In [41]:
data1.dtypes

Age                          int64
Attrition                   object
BusinessTravel              object
Department                  object
DistanceFromHome             int64
Education                    int64
EducationField              object
EmployeeCount                int64
EmployeeID                   int64
Gender                      object
JobLevel                     int64
JobRole                     object
MaritalStatus               object
MonthlyIncome                int64
NumCompaniesWorked         float64
Over18                      object
PercentSalaryHike            int64
StandardHours                int64
StockOptionLevel             int64
TotalWorkingYears          float64
TrainingTimesLastYear        int64
YearsAtCompany               int64
YearsSinceLastPromotion      int64
YearsWithCurrManager         int64
dtype: object

In [42]:
from sklearn.preprocessing import LabelEncoder
le = LabelEncoder()

In [43]:
data1['Attrition'] = le.fit_transform(data1['Attrition'])
data1['Gender'] = le.fit_transform(data1['Gender'])

In [44]:
data1.drop(['BusinessTravel','Department','EducationField','JobRole','MaritalStatus','Over18','EmployeeID','EmployeeCount'] , axis = 1 , inplace = True)
data1.dtypes

Age                          int64
Attrition                    int32
DistanceFromHome             int64
Education                    int64
Gender                       int32
JobLevel                     int64
MonthlyIncome                int64
NumCompaniesWorked         float64
PercentSalaryHike            int64
StandardHours                int64
StockOptionLevel             int64
TotalWorkingYears          float64
TrainingTimesLastYear        int64
YearsAtCompany               int64
YearsSinceLastPromotion      int64
YearsWithCurrManager         int64
dtype: object

In [56]:
y = data1.Attrition
X= data1[['Age','DistanceFromHome','Education','Gender', 'JobLevel', 'MonthlyIncome','NumCompaniesWorked', 'PercentSalaryHike', 'StandardHours','StockOptionLevel', 'TotalWorkingYears', 'TrainingTimesLastYear','YearsAtCompany', 'YearsSinceLastPromotion', 'YearsWithCurrManager']]
X1= sm.add_constant(X)

In [57]:
logist = sm.Logit(y,X1)

In [58]:
result2 = logist.fit()

Optimization terminated successfully.
         Current function value: 0.406815
         Iterations 7


In [60]:
result2.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                           Logit Regression Results                           
==============================================================================
Dep. Variable:              Attrition   No. Observations:                 4410
Model:                          Logit   Df Residuals:                     4395
Method:                           MLE   Df Model:                           14
Date:                Thu, 20 Aug 2020   Pseudo R-squ.:                 0.07897
Time:                        15:13:18   Log-Likelihood:                -1794.1
converged:                       True   LL-Null:                       -1947.9
Covariance Type:            nonrobust   LLR p-value:                 3.011e-57
===========================================================================================
                              coef    std err          z      P>|z|      [0.025      0.975]
-------------------------------------------------------------------------------------------
Age                        -0.0358      0.007     -5.299      0.000      -0.049      -0.023
DistanceFromHome           -0.0026      0.005     -0.491      0.623      -0.013       0.008
Education                  -0.0539      0.042     -1.296      0.195      -0.135       0.028
Gender                      0.0780      0.088      0.889      0.374      -0.094       0.250
JobLevel                   -0.0312      0.039     -0.801      0.423      -0.108       0.045
MonthlyIncome           -2.261e-06   9.39e-07     -2.409      0.016    -4.1e-06   -4.21e-07
NumCompaniesWorked          0.1082      0.018      5.996      0.000       0.073       0.144
PercentSalaryHike           0.0131      0.012      1.141      0.254      -0.009       0.036
StandardHours               0.0725      0.043      1.674      0.094      -0.012       0.157
StockOptionLevel           -0.0483      0.050     -0.956      0.339      -0.147       0.051
TotalWorkingYears          -0.0541      0.012     -4.622      0.000      -0.077      -0.031
TrainingTimesLastYear      -0.1323      0.035     -3.825      0.000      -0.200      -0.064
YearsAtCompany              0.0023      0.018      0.127      0.899      -0.033       0.038
YearsSinceLastPromotion     0.1230      0.020      6.103      0.000       0.084       0.163
YearsWithCurrManager       -0.1287      0.022     -5.948      0.000      -0.171      -0.086
===========================================================================================
"""

__By the above result summary we can conclude that the following features (Age,MonthlyIncome,NumCompaniesWorked,TotalWorkingYears,TrainingTimesLastYear,YearsSinceLastPromotion,YearsWithCurrManager) are have significant impact on dependent variable(Attrition)__  

__To find the probability of having Attrition or not, we have to calculate the 'P' value by placing the coefficient values into the formula.  
If 'P' value is > 0.5 , chance of having attrition is good.  
If 'P' value is < 0.5 , chance of having attrition is bad.__

__For prediction we are considering the following values as independent variables with coefficient values respectivetely:__  
  
Age                          25 (-0.0358)  
MonthlyIncome                60,000 (-2.261e-06)  
NumCompaniesWorked           2 (0.108)  
TotalWorkingYears            4 (-0.054)  
TrainingTimesLastYear        2 (-0.132)  
YearsSinceLastPromotion      1 (0.123)  
YearsWithCurrManager         2 (-0.128)  

__By following the same calculation of above problem, we got the 'P' value as:__  
  
P = 

In [68]:
b = 0-(0.0358*(25))-(0.000002261*(60000))+(0.108*(2))-(0.054*(4))-(0.132*(2))+(0.123*(1))-(0.128*(2))
b

-1.4276600000000002

In [69]:
p = 1/(1+(math.exp(-(b))))
p

0.19346354519406397

__P = 0.193  
As 'P' value is less than 0.5, the probability of attrition is very low.__